### PR TITLE
Fix failing test_invalid_input_csr_large

### PR DESCRIPTION
**Pull request by Flamefire**
The test introduced in #102530 has a bug:   
Construction of `crow_indices` raises an exception: "value cannot be converted to type int32 without overflow" which is obviously correct.
This makes the test fail which is supposed to check for an overflow in nnz.
Fix by making the construction of `crow_indices` pass although with an invalid value which would error later but triggers the correct check.

Given that I'm not sure it is even worth checking for an overflow in nnz:
- `crow_indices[..., -1] == nnz` is already enforced
- this can only hold if `crow_indices` is able to hold `nnz` without overflow
- `col_indices` has to be of the same type as `crow_indices`
- Hence the type of `col_indices` has to be able to hold the value of `nnz`

So in conclusion: The situation being checked for cannot reasonably occur

CC @pearu as the test author for additional insight

cc @alexsamardzic @nikitaved @pearu @cpuhrsch @amjames @bhosmer

### CODE REVIEW

File: test/test_sparse_csr.py [hunk 1]
```diff
@@ -866,7 +866,11 @@ def test_invalid_input_csr_large(self):
 
         nnz = 2 ** 31
         with self.assertRaisesRegex(RuntimeError, '32-bit integer overflow in nnz'):
-            torch.sparse_csr_tensor(torch.tensor([0, nnz // 2, nnz], dtype=torch.int32),
+            # nnz cannot be stored in int32 crow_indices
+            # but the `crow_indices[..., -1] == nnz`` check happens after the overflow validation
+            # So we can use `nnz - 1` here to avoid `value cannot be converted to type int32 without overflow`
+            # during construction of crow_indices
+            torch.sparse_csr_tensor(torch.tensor([0, nnz // 2, nnz - 1], dtype=torch.int32),
```
Comment (pearu) [inline test/test_sparse_csr.py:873]: The suggested fix is relies on the order of invariant checks. This is not ideal in the sense that when the order of invariant checks is changed for any sensible reason, the construction of a sparse CSR tensor may fail again. For instance, the suggested input to `sparse_csr_tensor` does not satisfy the invariant `col_indices.shape[-1] == nnz` nor `values.shape[0] == crow_indices[-1]`.

This test makes sense only when `sparse_csr_tensor` fails or succeeds depending on the dtype of `crow/col_indices`. For this reason, the `sparse_csr_tensor` call in the `with`-block and the call after the `with`-block ought to be identical except in `crow/col_indices` dtype.

Therefore, another possible fix is
```suggestion
        with self.assertRaisesRegex(RuntimeError, '32-bit integer overflow in nnz'):
            torch.sparse_csr_tensor(torch.tensor([0, nnz // 2, nnz], dtype=torch.int64).to(torch.int32),
```
or equivalently,
```suggestion
        with self.assertRaisesRegex(RuntimeError, '32-bit integer overflow in nnz'):
            torch.sparse_csr_tensor(torch.tensor([0, nnz // 2, -nnz], dtype=torch.int32),
```

Btw, when `crow_indices[-1] < 0`, the `sparse_csr_tensor` call always fails with `32-bit integer overflow in nnz` exception.

**Review by pearu — CHANGES_REQUESTED** [review]
@Flamefire, thanks for sending the fix! As discussed below, the suggested fix is a bit problematic because it uses inputs to sparse_csr_tensor that invalidate the sparse compressed tensor invariants in other ways.

An alternative fix is proposed below.

File: test/test_sparse_csr.py (hunk 1 above)
Comment (Flamefire) [reply test/test_sparse_csr.py:873]: It looks to me like your suggested change also depends on the order of invariant checks. Using `torch.tensor([0, nnz // 2, nnz], dtype=torch.int64).to(torch.int32)` relies on undefined behavior (prior to C++20) and PyTorch missing to check for overflow in the `to` conversion which it does in the construction but let's assume it to be equivalent to `torch.tensor([0, nnz // 2, -nnz], dtype=torch.int32)`.   
Then `crow_indices[..., -1] == nnz` is not satisfied (which you mentioned as `values.shape[0] == crow_indices[-1]` which is equivalent but the former is what is checked in code) as `-nnz != nnz`   

I don't understand the comment about `col_indices.shape[-1] == nnz`. `nnz` is the number of values: `const int64_t values_nnz = values.size(batch_ndim);`, i.e. `torch.ones(nnz, dtype=torch.int8).size()` and `col_indices = torch.arange(nnz // 2, dtype=torch.int32).repeat(2)`. So that invariant holds, doesn't it?

> Btw, when `crow_indices[-1] < 0`, the `sparse_csr_tensor` call always fails with `32-bit integer overflow in nnz` exception.

That is the case for any `crow_indices[-1]` when it is a 32bit tensor and `values.size()` overflows that. Again depending on the order of the invariant checks:

```
import torch
torch.sparse.check_sparse_tensor_invariants.enable()
torch.sparse_csr_tensor(torch.tensor([0, 2, 4], dtype=torch.int64),torch.arange(2, dtype=torch.int64).repeat(2),torch.ones(4, dtype=torch.int8), (2, 1024 // 2)) # OK
torch.sparse_csr_tensor(torch.tensor([0, 2, -4], dtype=torch.int64),torch.arange(2, dtype=torch.int64).repeat(2),torch.ones(4, dtype=torch.int8), (2, 1024 // 2)) # `crow_indices[..., -1] == nnz` is not satisfied
nnz = 2 ** 31
col_indices = torch.arange(nnz // 2, dtype=torch.int32).repeat(2)
values = torch.ones(nnz, dtype=torch.int8)
for n in (-1, 0, 1, 5):
  torch.sparse_csr_tensor(torch.tensor([0, nnz // 2, n], dtype=torch.int32), col_indices, values, (2, nnz // 2)) # 32-bit integer overflow in nnz = 2147483648
```

So IMO my solution is already as good as possible: If that test fails because the invariant check order has changed the comment clearly explains that. Your suggestion fail the same invariant checks and if PyTorch starts checking for overflow in `tensor.to` it would additionally fail with a "mysterious" error as seen currently where the failing test is seemingly disabled on CI as nobody understood the failure reason. (BTW: Where can I see that this test is skipped on CI? I've only seen it in the comment at https://github.com/pytorch/pytorch/issues/102673)

The difference now is: `torch.tensor([0, nnz // 2, nnz - 1], dtype=torch.int32)` vs `torch.tensor([0, nnz // 2, nnz], dtype=torch.int64)` which is IMO similar enough. The alternative of `torch.tensor([0, nnz // 2, -nnz], dtype=torch.int32)` or `torch.tensor([0, nnz // 2, nnz], dtype=torch.int64).to(torch.int32)` IMO just make that constraint violation less clear.

Comment (pearu) [reply test/test_sparse_csr.py:873]: I agree with the point re undefined behavior.

> I don't understand the comment about col_indices.shape[-1] == nnz

I can see that. The requirement is that all three quantities --- `col_indices.shape[-1]`, `values.shape[0]`, and `crow_indices[-1]` --- must be equal. The definition of `nnz` is kind of arbitrary. In code, it is defined as `values.shape[0]` while I was considering if defined as `crow_indices[-1]`.

> Where can I see that this test is skipped on CI?

I believe the test is skipped on CI because the test method is decorated with `largeTensorTest("30GB", "cpu")`.

> So IMO my solution is already as good as possible

I disagree because from just reading the code, test sample is clearly invalid because the invariant `col_indices.shape[-1] == values.shape[0] == crow_indices[-1]` does not hold and I would expect that an exception is raised about the invariant being invalidated, not a signed integer overflow exception.

The original indent of the test was to cover the case where the int64->int32 conversion will silently lead to negative entries in `crow_indices` that cause hard-to-debug crashes by algorithms processing `crow_indices` that last item may or may not be equal to `values.shape[0]`.

Btw, the int64->int32 conversion does not only happen in the `test_invalid_input_csr_large` method but also in various calls to sparse library functions that support only 32-bit indices but users have provided a CSR tensor with 64-bit indices. Unfortunately, these cases are currently uncovered because signed integer overflows are not detected in `torch.Tensor.to`.

Notice that when torch.Tensor.to will implement the signed integer overflow check, the corresponding exception will be raised during the int64->int32 conversion. Here we are testing the exception raised by the `_validate_compressed_sparse_indices` function on the case where `values.shape[-1]` (that we define as `nnz`) cannot be safely converted to a `dtype` of the `crow/col_indices`. Invariant checks like `crow_indices[-1] == nnz` cannot be performed before we know that `nnz` has a valid value when converted to `crow_indices.dtype`. That said, I now I agree that your solution is technically fine and I don't have a solution to my disagreement above.

**Review by pearu — APPROVED** [review]
Thanks!

File: test/test_sparse_csr.py (hunk 1 above)
Comment (Flamefire) [reply test/test_sparse_csr.py:873]: > I agree with the point re undefined behavior.

Looks like I need to correct myself and the (implicit) cast happening there is implementation defined behavior, not undefined. But still worrisome.

> I disagree because from just reading the code, test sample is clearly invalid because the invariant `col_indices.shape[-1] == values.shape[0] == crow_indices[-1]` does not hold and I would expect that an exception is raised about the invariant being invalidated, not a signed integer overflow exception.

Yes but that's what the comment is for: Explaining the reasoning and expected outcome for why it is that way and tests what it is supposed to. Technically I like your `to(int32)` better but it would need a similar, likely even more confusing comment explaining what happens there and why that might fail the (same) invariant check.

> The original indent of the test was to cover the case where the int64->int32 conversion will silently lead to negative entries in `crow_indices` that cause hard-to-debug crashes by algorithms processing `crow_indices` that last item may or may not be equal to `values.shape[0]`.

Understood. I assume the overflow check on construction was added after this test.

> Unfortunately, these cases are currently uncovered because signed integer overflows are not detected in `torch.Tensor.to`.

I'd guess due to the compute overhead involved in adding those checks. Still might cause quite some pain when values suddenly change considerably.

> Notice that when torch.Tensor.to will implement the signed integer overflow check, the corresponding exception will be raised during the int64->int32 conversion. Here we are testing the exception raised by the `_validate_compressed_sparse_indices` function on the case where `values.shape[-1]` (that we define as `nnz`) cannot be safely converted to a `dtype` of the `crow/col_indices`. Invariant checks like `crow_indices[-1] == nnz` cannot be performed before we know that `nnz` has a valid value when converted to `crow_indices.dtype`. That said, I now I agree that your solution is technically fine and I don't have a solution to my disagreement above.

Yes, that's what I concluded earlier: When `nnz` overflows the type of `crow_indices` it cannot possibly satisfy the invariant. Or can it? The implementation defined behavior is usually a `mod 2N`. So in our case `crow_indices[-1] = 0` and  `values.shape[-1] = 2^32` would overflow but satisfy the constraint (after the overflow).   
So the check is required and I'd say it is ok to violate one constraint given it is clearly documented that we use knowledge of the implementation here.

BTW: While talking about this: The check for index overflow is done using that implementation defined behavior (it basically checks if the cast changes the value). It might better be replaced by the `overflows` utility function call used e.g. during construction of a tensor form a python numeric value.

**Review by cpuhrsch — APPROVED** [review]
